pkg/estimator/client/interface.go

```diff
@@ -30,26 +32,83 @@ import (
 const UnauthenticReplica = -1
 
 var (
-	replicaEstimators              = map[string]ReplicaEstimator{}
-	unschedulableReplicaEstimators = map[string]UnschedulableReplicaEstimator{}
+	// replicaEstimators are organized in a TreeMap, sorted by descending priority.
+	// The key is of type EstimatorPriority, indicating the estimator's priority level.
+	// The value is a map with string keys and ReplicaEstimator values, grouping estimators by their respective priorities.
+	replicaEstimators = treemap.NewWith(estimatorPriorityComparator)
+
+	// unschedulableReplicaEstimators are organized in a TreeMap, sorted by descending priority.
+	// The key is of type EstimatorPriority, indicating the estimator's priority level.
+	// The value is a map with string keys and UnschedulableReplicaEstimator values, grouping estimators by their respective priorities.
+	unschedulableReplicaEstimators = treemap.NewWith(estimatorPriorityComparator)
 )
 
+// registerReplicaEstimator add a estimator to replicaEstimators
+func registerReplicaEstimator(estimatorName string, estimator ReplicaEstimator) {
+	if val, ok := replicaEstimators.Get(estimator.Priority()); !ok {
+		replicaEstimators.Put(estimator.Priority(), map[string]ReplicaEstimator{estimatorName: estimator})
```

Comment:
Is it necessary to use `map[string]ReplicaEstimator` here? I see that the key of map is ignored when using it. Maybe the slice  `[]ReplicaEstimator` is enough.

Reply:
No, this can prevent the `RegisterSchedulerEstimator` from being called repeatedly. 

One obvious issue is, the test function `TestDescheduler_worker` will fail since it call `RegisterSchedulerEstimator` for many times, if slice used, there will be redundant and dirty estimators, while if map used, it can help with deduplication. 

This is also why `UnRegisterSchedulerEstimator ` be added in my previous commit, since the map is introduced, the `UnRegisterSchedulerEstimator` is needless.

Reply:
We don’t have to compromise code implementation because of the test function `TestDescheduler_worker` but modify the test code.
```go
        // register estimator outside the loop.
        estimatorCache := estimatorclient.NewSchedulerEstimatorCache()
	schedulerEstimator := estimatorclient.NewSchedulerEstimator(estimatorCache, 5*time.Second)
	estimatorclient.RegisterSchedulerEstimator(schedulerEstimator)

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			ctx, cancel := context.WithCancel(context.TODO())
			defer cancel()

			binding, err := buildBinding(tt.args.name, tt.args.namespace, tt.args.target, tt.args.status)
			if err != nil {
				t.Fatalf("build binding error: %v", err)
			}

			karmadaClient := fakekarmadaclient.NewSimpleClientset(binding)
			factory := informerfactory.NewSharedInformerFactory(karmadaClient, 0)

			desched := &Descheduler{
				KarmadaClient:           karmadaClient,
				informerFactory:         factory,
				bindingInformer:         factory.Work().V1alpha2().ResourceBindings().Informer(),
				bindingLister:           factory.Work().V1alpha2().ResourceBindings().Lister(),
				clusterInformer:         factory.Cluster().V1alpha1().Clusters().Informer(),
				clusterLister:           factory.Cluster().V1alpha1().Clusters().Lister(),
				schedulerEstimatorCache: estimatorCache,
				unschedulableThreshold:  5 * time.Minute,
				eventRecorder:           record.NewFakeRecorder(1024),
			}
```

Reply:
I haven't closely examined whether the test code you provided works, but the essence of the problem isn't about compromising with the test code. The test code merely exposes this issue, what I expect is that even in normal code, issues of duplicate registration will also be avoided.
Besides, the advantages of slice over map are not that significant.